content/documentation/internals/nats-streaming-protocol.md

```diff
@@ -0,0 +1,199 @@
+  parent = "Internals"
++++
+
+# NATS streaming protocol
+
+The NATS streaming protocol sits atop the [core NATS protocol](../nats-protocol) and uses [Google's protocol buffers](https://developers.google.com/protocol-buffers/).  Protocol buffer messages are marshalled into bytes and published as core NATS messages on specific subjects described below.  In communicating with the NATS streaming server, The NATS [request/reply](../../concepts/nats-req-rep) pattern is used for all protocol messages that have a corresponding reply.
+
+## NATS streaming protocol conventions
+
+**Subject names**: Subject names, including reply subject (INBOX) names, are case-sensitive and must be non-empty alphanumeric strings with no embedded whitespace, and optionally token-delimited using the dot character (`.`), e.g.:
+
+`FOO`, `BAR`, `foo.bar`, `foo.BAR`, `FOO.BAR` and `FOO.BAR.BAZ` are all valid subject names
+
+`FOO. BAR`, `foo. .bar` and`foo..bar` are *not* valid subject names
+
+**Wildcards**: NATS streaming does *not* support wildcards in subject subscriptions.
+
+**Protocol definition**: The fields of NATS streaming protocol messages are defined in the go-nats-streaming [protocol file](https://github.com/nats-io/go-nats-streaming/blob/master/pb/protocol.proto).
+
+
+## NATS streaming protocol messages
+
+The following table briefly describes the NATS streaming protocol messages.
+
+Click the name to see more detailed information, including usage:
+
+| Message Name                      | Sent By |    Description
+| --------------------------------- |:--------|:--------------------------------------------
+| [`ConnectRequest`](#CONNREQ)      | Client  | Request to connect to the streaming server
+| [`ConnectResponse`](#CONNRESP)    | Server  | Result of a connection request
+| [`SubscriptionRequest`](#SUBREQ)  | Client  | Request sent to Subscribe to a subject and retrieve data
+| [`SubscriptionResponse`](#SUBRESP)| Server  | Result of a subscription request
+| [`UnsubscribeRequest`](#UNSUBREQ) | Client  | Unsubscribe from a subject
+| [`PubMsg`](#PUBMSG)               | Client  | Publish a message to a subject, with optional reply subject
+| [`PubAck`](#PUBACK)               | Server  | An acknowledgement that a published message has been persisted
+| [`MsgProto`](#MSGPROTO)           | Server  | A message (originally published from a client) to a subscribing client
+| [`Ack`](#ACK)                     | Client  | Acknowledges that a message has been received
+| [`CloseRequest`](#CLOSEREQ)       | Client  | Request sent to close the connection to the streaming server
+| [`CloseResp`](#CLOSERESP)         | Server  | Result of the close request
+
+The following sections explain each protocol message.
+
+## <a name="CONNREQ"></a>ConnectRequest
+
+### Message Structure
+
+* `ClientID`: A unique identifier for a client.
+* `HeartbeatInbox`: An inbox to which the streaming server will send heartbeats to for the client to process.
```

Comment:
Looks like we do that for the other types. So the only thing incorrect here is that `ClientID` and `HeartbeatInbox` should be `clientID` and `heartbeatInbox` instead.